UI Select does not pull all data

tazo90 opened the pull request:
Ref: [https://github.com/quantmind/lux/issues/158](https://github.com/quantmind/lux/issues/158).

lsbardel commented:
@tazo90 can you make sure the pull request can be merged with master?

tazo90 commented:
@lsbardel sure, let me check.

reupen commented on luxjs/forms/utils.js:
```diff
@@ -3,54 +3,290 @@
-            api.get(null, params).then(function (data) {
-                if (attrs.multiple) {
-                    options.splice(0, 1);
+            remoteService.query(api, target, scope, attrs, config, null, false);
```
Here we are calling `remoteService.query` with `null` for the `searchValue` argument, but in the function itself it's comparing `searchValue` with `undefined`.

reupen commented on luxjs/forms/utils.js:
```diff
@@ -3,54 +3,290 @@
+                    config.optionsTotal = data.data.total;
+
+                    if (searchValue !== undefined && data.data.result.length === 0) {
+                        options[0].name = 'Cannot find value';
```
If there are no search results, the message should something like `No matches found`.

reupen commented:
Hi @tazo90 

Just a few other things I spotted (on single-select selects).

If you scroll to the end of the list quickly, it will request an offset past the last item. It is also incorrectly replacing the `Please select...` with `Cannot find value` when this happens (this is related to one of the comments above). I haven't looked in detail why it is requesting past the last item but we should be careful to make sure a request for more items isn't started if one is already in progress.

Another thing is, if you're on an update form and the current value of the select field is past the first page of data, when you open the select drop down list it will have `Please select...` selected. Not sure what the best solution here is.

Thanks

reupen commented on luxjs/forms/utils.js:
```diff
@@ -3,54 +3,290 @@
+    }])
+    //
+    .directive('selectInfinity', ['$parse', '$timeout', function($parse, $timeout) {
+        function height(elem) {
```
Is this the best place for these functions? Can we write unit tests for them if they're here?

reupen commented:
Also, how did it go adding tests for the `selectInfinity` directive?

tazo90 commented:
@reupen thanks for review, I'll make the necessary changes.

tazo90 commented:
I made updates against master. Tests on travis are ok now.

lsbardel commented:
Hi, this is a chunky pull request.
Can I ask you to add comments to the various directives and methods?
And a release note in the [release/note.md](https://github.com/quantmind/lux/blob/master/release/notes.md) file?

tazo90 commented:
@lsbardel sure, I will make required updates.

tazo90 commented:
I added more comments and updated the `release/notes.md`.

In order to test this fix please comment [line](https://github.com/quantmind/lux/blob/master/luxjs/forms/form.js#L456) and you can check behaviour on edit page user.

reupen commented:
Hi @tazo90 

Sorry, could you update against master again?

Thanks.

tazo90 commented:
@reupen PR has been updated.

reupen commented:
Hi @tazo90,

For a multi-select field, if there are initial selected items outside the first page, they aren't displayed properly:

![groups](https://cloud.githubusercontent.com/assets/12693549/11779519/82cd56ae-a253-11e5-86f2-87ff1c280c1a.png)

The API endpoint returns current values in the form:
```
[{

    "id": ​28,
    "repr": "orgabc-admin"

},
{

    "id": ​29,
    "repr": "staff-admin"

}]
```

Can we use the `repr` field (if present) from this to fix it?

reupen commented:
When entering search text in a multi-select field, the search box is a bit small and the text gets cut off:

![multi-search](https://cloud.githubusercontent.com/assets/12693549/11780191/1100ab5c-a258-11e5-9903-bf592b5b6191.png)

Is there anything we can do to improve that?

reupen commented on lux/media/lux/lux.js:
```diff
@@ -3430,58 +3447,427 @@ angular.module('lux.form.process', ['ngFileUpload'])
+                        }
+                    });
+
+                    defer.resolve(data);
```
Should this be resolved here with the require block above? If yes, could we add a comment?

reupen commented on lux/media/lux/lux.js:
```diff
@@ -3430,58 +3447,427 @@ angular.module('lux.form.process', ['ngFileUpload'])
+         */
+        function height(elem) {
+            elem = elem[0] || elem;
+            if (isNaN(elem.offsetHeight)) {
```
What is the isNaN check here for? isNaN has odd results because it coerces the argument to a number first. 
When is `elem.document` defined?

reupen commented on lux/media/lux/lux.js:
```diff
@@ -3430,58 +3447,427 @@ angular.module('lux.form.process', ['ngFileUpload'])
+        function pageYOffset(elem) {
+            elem = elem[0] || elem;
+            if (isNaN(window.pageYOffset)) {
+                return elem.document.documentElement.scrollTop;
```
Similar question as the previous one here - what is this logic here intended to do? When is `elem.document` defined? Please add comments.
